README.md
web interface for graphite
=======
## Overview

Graphite consists of two major components:

1. the frontend Django webapp that runs under Apache/mod_wsgi
2. the backend carbon-cache.py daemon

Client applications connect to the running carbon-cache.py daemon on port 2003 and send it
lines of text of the following format:

    my.metric.name value unix_timestamp

For example:

    performance.servers.www01.cpuUsage 42.5 1208815315

The metric name is like a filesystem path that uses `.` as a separator instead of `/`.
The value is some scalar integer or floating point value.
The unix_timestamp is unix epoch time, as an integer.

Each line like this corresponds to one data point for one metric.

Once you've got some clients sending data to carbon-cache, you can view
graphs of that data in the frontend webapp.


## Webapp Installation

Use the instructions in the INSTALL file.


## Running carbon-cache.py

First you must tell carbon-cache what user it should run as.
This must be a user with write privileges to `$GRAPHITE_ROOT/storage/whisper/`
Specify the user account in `$GRAPHITE_ROOT/carbon/conf/carbon.conf`

This user must also have write privileges to `$GRAPHITE_ROOT/storage/log/carbon-cache/`


## Writing a client

First you obviously need to decide what data it is you want to graph with
graphite. The script `examples/example-client.py` demonstrates a simple client
that sends loadavg data for your local machine to carbon on a minutely basis.

The default storage schema stores data in one-minute intervals for 2 hours.
This is probably not what you want so you should create a custom storage schema
according to the docs on the graphite wiki (http://graphite.wikidot.com).


debian/changelog
yabs-graphite-ceres (0.2.2.6) unstable; urgency=low

  * another one rebuild 

 -- Kulikovsky Dmitry <dkulikovsky@yandex-team.ru>  Fri, 28 Feb 2014 18:23:41 +0400

yabs-graphite-ceres (0.2.2.5) unstable; urgency=low

  * fixed init script again 

 -- Kulikovsky Dmitry <dkulikovsky@yandex-team.ru>  Fri, 28 Feb 2014 15:40:13 +0400

yabs-graphite-ceres (0.2.2.2) unstable; urgency=low

  * init script moved to django-admin 

 -- Kulikovsky Dmitry <dkulikovsky@yandex-team.ru>  Fri, 28 Feb 2014 15:27:32 +0400

yabs-graphite-ceres (0.2.2.1) unstable; urgency=low

  * fixed few upstream bugs 

 -- Kulikovsky Dmitry <dkulikovsky@yandex-team.ru>  Fri, 28 Feb 2014 15:24:13 +0400

yabs-graphite-ceres (0.2.2.0) unstable; urgency=low

  * merged with upstream 

 -- Kulikovsky Dmitry <dkulikovsky@yandex-team.ru>  Thu, 27 Feb 2014 19:02:38 +0400

yabs-graphite-ceres (0.2.1.4) unstable; urgency=low

  * applied patch
  https://github.com/cbowman0/graphite-web/commit/b529ff647143024f0c650c1b153a9d6c5ea27f70
 

 -- Kulikovsky Dmitry <dkulikovsky@yandex-team.ru>  Mon, 17 Feb 2014 18:59:28 +0400

yabs-graphite-ceres (0.2.1.3) unstable; urgency=low

  * Adding discrete points sticking function 

 -- Anna Sysoeva <antrue@yandex-team.ru>  Mon, 20 Jan 2014 15:46:10 +0400

yabs-graphite-ceres (0.2.1.2) unstable; urgency=low

  * added event_by_date 

 -- Kulikovsky Dmitry <dkulikovsky@yandex-team.ru>  Fri, 17 Jan 2014 20:11:57 +0400

yabs-graphite-ceres (0.2.1.1) unstable; urgency=low

  * reduced size of fonts in menu

 -- Kulikovsky Dmitry <dkulikovsky@yandex-team.ru>  Thu, 09 Jan 2014 15:43:33 +0400

yabs-graphite-ceres (0.2.1) unstable; urgency=low

  * version collision, rebuild

 -- Kulikovsky Dmitry <dkulikovsky@yandex-team.ru>  Thu, 09 Jan 2014 14:10:39 +0400

yabs-graphite-ceres (0.2.0.28) unstable; urgency=low

  * reduced size of * in dashboard

 -- Kulikovsky Dmitry <dkulikovsky@yandex-team.ru>  Thu, 09 Jan 2014 13:21:17 +0400

yabs-graphite-ceres (0.2.0.27) unstable; urgency=low

  * removed nginx dep

 -- Kulikovsky Dmitry <dkulikovsky@yandex-team.ru>  Fri, 06 Dec 2013 20:06:25 +0400

yabs-graphite-ceres (0.2.0.26) unstable; urgency=low

  * removed useless files in opt

 -- Kulikovsky Dmitry <dkulikovsky@yandex-team.ru>  Fri, 06 Dec 2013 13:57:22 +0400

yabs-graphite-ceres (0.2.0.25) precise; urgency=low

  * added construct button in dashboard

 -- Kulikovsky Dmitry <dkulikovsky@yandex-team.ru>  Wed, 04 Dec 2013 23:38:20 +0400

yabs-graphite-ceres (0.2.0.24) unstable; urgency=low

  * moved to github
  * few minor fixes 

 -- Kulikovsky Dmitry <dkulikovsky@yandex-team.ru>  Wed, 04 Dec 2013 21:45:56 +0400

yabs-graphite-ceres (0.2.0.23) unstable; urgency=low

  * depends fix

 -- Anna Sysoeva <antrue@yandex-team.ru>  Wed, 20 Nov 2013 21:11:16 +0400

yabs-graphite-ceres (0.2.0.22) unstable; urgency=low

  * depends fix

 -- Anna Sysoeva <antrue@yandex-team.ru>  Wed, 20 Nov 2013 20:59:24 +0400

yabs-graphite-ceres (0.2.0.21) unstable; urgency=low

  * django versions fix

 -- Anna Sysoeva <antrue@yandex-team.ru>  Wed, 20 Nov 2013 20:55:57 +0400

yabs-graphite-ceres (0.2.0.20) unstable; urgency=low

  * django versions fix

 -- Anna Sysoeva <antrue@yandex-team.ru>  Wed, 20 Nov 2013 20:50:09 +0400

yabs-graphite-ceres (0.2.0.19) unstable; urgency=low

  * depends on django fixed

 -- Anna Sysoeva <antrue@yandex-team.ru>  Wed, 20 Nov 2013 20:32:12 +0400

yabs-graphite-ceres (0.2.0.18) unstable; urgency=low

  * depends fixed 

 -- Anna Sysoeva <antrue@yandex-team.ru>  Wed, 20 Nov 2013 20:27:56 +0400

yabs-graphite-ceres (0.2.0.17) unstable; urgency=low

  * logrotate removed, moved to yabs-graphite-ceres-config 

 -- Anna Sysoeva <antrue@yandex-team.ru>  Wed, 20 Nov 2013 14:42:51 +0400

yabs-graphite-ceres (0.2.0.16) unstable; urgency=low

  * rollup logging switched off

 -- Anna Sysoeva <antrue@yandex-team.ru>  Wed, 06 Nov 2013 18:33:44 +0400

yabs-graphite-ceres (0.2.0.15) precise; urgency=low

  * read metadata only once in rollup
  * print broken metrics (broken .ceres-node json) to log 

 -- Kirill Zhalnin <kiwi@yandex-team.ru>  Fri, 15 Nov 2013 18:47:24 +0400

yabs-graphite-ceres (0.2.0.14) precise; urgency=low

  * removed bad depend for python-django-1.5 

 -- Kulikovsky Dmitry <dkulikovsky@yandex-team.ru>  Wed, 13 Nov 2013 14:36:01 +0400

yabs-graphite-ceres (0.2.0.13) precise; urgency=low

  * fixed shorcuts+zoom bug
  * made djandgo depend more universal

 -- Kulikovsky Dmitry <dkulikovsky@yandex-team.ru>  Wed, 13 Nov 2013 14:28:23 +0400

yabs-graphite-ceres (0.2.0.12) precise; urgency=low

  * rollup stat to ceres_rollup.err 

 -- Kirill Zhalnin <kiwi@yandex-team.ru>  Thu, 07 Nov 2013 21:00:33 +0400

yabs-graphite-ceres (0.2.0.11) precise; urgency=low

  * rollup logrotate 

 -- Kirill Zhalnin <kiwi@yandex-team.ru>  Thu, 07 Nov 2013 17:02:52 +0400

yabs-graphite-ceres (0.2.0.10) precise; urgency=low

  * new rollup

 -- Kirill Zhalnin <kiwi@yandex-team.ru>  Thu, 07 Nov 2013 16:40:22 +0400

yabs-graphite-ceres (0.2.0.9) precise; urgency=low

  * fixed dashboards links

 -- Kulikovsky Dmitry <dkulikovsky@yandex-team.ru>  Wed, 06 Nov 2013 17:39:13 +0400

yabs-graphite-ceres (0.2.0.8) unstable; urgency=low

  * added shortcuts for timeRange periods

 -- Kulikovsky Dmitry <dkulikovsky@yandex-team.ru>  Wed, 06 Nov 2013 15:14:06 +0400

yabs-graphite-ceres (0.2.0.7) unstable; urgency=low

  * fixed series name output for diffSeries

 -- Kulikovsky Dmitry <dkulikovsky@yandex-team.ru>  Wed, 30 Oct 2013 18:16:52 +0400

yabs-graphite-ceres (0.2.0.6) unstable; urgency=low

  * fixed sort order in finder

 -- Kulikovsky Dmitry <dkulikovsky@yandex-team.ru>  Tue, 29 Oct 2013 13:48:43 +0400

>>>>>>> .r7382
yabs-graphite-ceres (0.2.0.5) unstable; urgency=low

  * merged to trunk 

 -- Kulikovsky Dmitry <dkulikovsky@yandex-team.ru>  Mon, 28 Oct 2013 15:03:24 +0400

yabs-graphite-ceres (0.2.0.4) unstable; urgency=low

  * fixed icons place 


webapp/graphite/settings.py
"""Copyright 2008 Orbitz WorldWide

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License."""
# Django settings for graphite project.
# DO NOT MODIFY THIS FILE DIRECTLY - use local_settings.py instead
import sys, os
from os.path import abspath, dirname, join
from warnings import warn


GRAPHITE_WEB_APP_SETTINGS_LOADED = False
WEBAPP_VERSION = '0.10.0-alpha'
DEBUG = False
JAVASCRIPT_DEBUG = False

# Filesystem layout
WEB_DIR = dirname( abspath(__file__) )
WEBAPP_DIR = dirname(WEB_DIR)
GRAPHITE_ROOT = dirname(WEBAPP_DIR)
# Initialize additional path variables
# Defaults for these are set after local_settings is imported
CONTENT_DIR = ''
CSS_DIR = ''
CONF_DIR = ''
DASHBOARD_CONF = ''
GRAPHTEMPLATES_CONF = ''
STORAGE_DIR = ''
WHITELIST_FILE = ''
INDEX_FILE = ''
LOG_DIR = ''
CERES_DIR = ''
WHISPER_DIR = ''
RRD_DIR = ''
STANDARD_DIRS = []

CLUSTER_SERVERS = []

# Cluster settings
CLUSTER_SERVERS = []
REMOTE_FIND_TIMEOUT = 3.0
REMOTE_FETCH_TIMEOUT = 6.0
REMOTE_RETRY_DELAY = 60.0
REMOTE_READER_CACHE_SIZE_LIMIT = 1000
CARBONLINK_HOSTS = ["127.0.0.1:7002"]
CARBONLINK_TIMEOUT = 1.0
CARBONLINK_HASHING_KEYFUNC = None
CARBONLINK_RETRY_DELAY = 15
REPLICATION_FACTOR = 1
MEMCACHE_HOSTS = []
FIND_CACHE_DURATION = 300
FIND_TOLERANCE = 2 * FIND_CACHE_DURATION
DEFAULT_CACHE_DURATION = 60 #metric data and graphs are cached for one minute by default
LOG_CACHE_PERFORMANCE = False

MAX_FETCH_RETRIES = 2

#Remote rendering settings
REMOTE_RENDERING = False #if True, rendering is delegated to RENDERING_HOSTS
RENDERING_HOSTS = []
REMOTE_RENDER_CONNECT_TIMEOUT = 1.0
LOG_RENDERING_PERFORMANCE = False

#Miscellaneous settings
SMTP_SERVER = "localhost"
DOCUMENTATION_URL = "http://graphite.readthedocs.org/"
ALLOW_ANONYMOUS_CLI = True
LOG_METRIC_ACCESS = False
LEGEND_MAX_ITEMS = 10
RRD_CF = 'AVERAGE'
STORAGE_FINDERS = (
    'graphite.finders.CeresFinder',
    'graphite.finders.StandardFinder',
)

#Authentication settings
USE_LDAP_AUTH = False
LDAP_SERVER = "" # "ldapserver.mydomain.com"
LDAP_PORT = 389
LDAP_USE_TLS = False
LDAP_SEARCH_BASE = "" # "OU=users,DC=mydomain,DC=com"
LDAP_BASE_USER = "" # "CN=some_readonly_account,DC=mydomain,DC=com"
LDAP_BASE_PASS = "" # "my_password"
LDAP_USER_QUERY = "" # "(username=%s)"  For Active Directory use "(sAMAccountName=%s)"
LDAP_URI = None

#Set this to True to delegate authentication to the web server
USE_REMOTE_USER_AUTHENTICATION = False

# Django 1.5 requires this so we set a default but warn the user
SECRET_KEY = 'UNSAFE_DEFAULT'

# Django 1.5 requires this to be set. Here we default to prior behavior and allow all
ALLOWED_HOSTS = [ '*' ]

# Override to link a different URL for login (e.g. for django_openid_auth)
LOGIN_URL = '/account/login'

# Set to True to require authentication to save or delete dashboards
DASHBOARD_REQUIRE_AUTHENTICATION = False
# Require Django change/delete permissions to save or delete dashboards.
# NOTE: Requires DASHBOARD_REQUIRE_AUTHENTICATION to be set
DASHBOARD_REQUIRE_PERMISSIONS = False
# Name of a group to which the user must belong to save or delete dashboards.  Alternative to
# DASHBOARD_REQUIRE_PERMISSIONS, particularly useful when using only LDAP (without Admin app)
# NOTE: Requires DASHBOARD_REQUIRE_AUTHENTICATION to be set
DASHBOARD_REQUIRE_EDIT_GROUP = None

DATABASES = {
  'default': {
    'NAME': '/opt/graphite/storage/graphite.db',
    'ENGINE': 'django.db.backends.sqlite3',
    'USER': '',
    'PASSWORD': '',
    'HOST': '',
    'PORT': '',
  },
}

# If using rrdcached, set to the address or socket of the daemon
FLUSHRRDCACHED = ''

## Load our local_settings
try:
  from graphite.local_settings import *
except ImportError:
  print >> sys.stderr, "Could not import graphite.local_settings, using defaults!"

## Load Django settings if they werent picked up in local_settings
if not GRAPHITE_WEB_APP_SETTINGS_LOADED:
  from graphite.app_settings import *

## Set config dependent on flags set in local_settings
# Path configuration
if not CONTENT_DIR:
  CONTENT_DIR = join(WEBAPP_DIR, 'content')
if not CSS_DIR:
  CSS_DIR = join(CONTENT_DIR, 'css')

if not CONF_DIR:
  CONF_DIR = os.environ.get('GRAPHITE_CONF_DIR', join(GRAPHITE_ROOT, 'conf'))
if not DASHBOARD_CONF:
  DASHBOARD_CONF = join(CONF_DIR, 'dashboard.conf')
if not GRAPHTEMPLATES_CONF:
  GRAPHTEMPLATES_CONF = join(CONF_DIR, 'graphTemplates.conf')

if not STORAGE_DIR:
  STORAGE_DIR = os.environ.get('GRAPHITE_STORAGE_DIR', join(GRAPHITE_ROOT, 'storage'))
if not WHITELIST_FILE:
  WHITELIST_FILE = join(STORAGE_DIR, 'lists', 'whitelist')
if not INDEX_FILE:
  INDEX_FILE = join(STORAGE_DIR, 'index')
if not LOG_DIR:
  LOG_DIR = join(STORAGE_DIR, 'log', 'webapp')
if not WHISPER_DIR:
  WHISPER_DIR = join(STORAGE_DIR, 'whisper/')
if not CERES_DIR:
  CERES_DIR = join(STORAGE_DIR, 'ceres/')
if not RRD_DIR:
  RRD_DIR = join(STORAGE_DIR, 'rrd/')
if not STANDARD_DIRS:
  try:
    import whisper
    if os.path.exists(WHISPER_DIR):
      STANDARD_DIRS.append(WHISPER_DIR)
  except ImportError:
    print >> sys.stderr, "WARNING: whisper module could not be loaded, whisper support disabled"
  try:
    import rrdtool
    if os.path.exists(RRD_DIR):
      STANDARD_DIRS.append(RRD_DIR)
  except ImportError:
    pass

# Default sqlite db file
# This is set here so that a user-set STORAGE_DIR is available
if 'sqlite3' in DATABASES.get('default',{}).get('ENGINE','') \
    and not DATABASES.get('default',{}).get('NAME'):
  DATABASES['default']['NAME'] = join(STORAGE_DIR, 'graphite.db')

# Caching shortcuts
if MEMCACHE_HOSTS:
  CACHE_BACKEND = 'memcached://' + ';'.join(MEMCACHE_HOSTS) + ('/?timeout=%d' % DEFAULT_CACHE_DURATION)

# Authentication shortcuts
if USE_LDAP_AUTH and LDAP_URI is None:
  LDAP_URI = "ldap://%s:%d/" % (LDAP_SERVER, LDAP_PORT)

if USE_REMOTE_USER_AUTHENTICATION:
  MIDDLEWARE_CLASSES += ('django.contrib.auth.middleware.RemoteUserMiddleware',)
  AUTHENTICATION_BACKENDS.insert(0,'django.contrib.auth.backends.RemoteUserBackend')

if USE_LDAP_AUTH:
  AUTHENTICATION_BACKENDS.insert(0,'graphite.account.ldapBackend.LDAPBackend')

if SECRET_KEY == 'UNSAFE_DEFAULT':
  warn('SECRET_KEY is set to an unsafe default. This should be set in local_settings.py for better security')
